### ISSUE TITLE
Variables declared only to pass to a closure are detected as unused

### ISSUE THREAD

**Issue by FractalizeR** [phpmd/phpmd#187]
``` php
    /**
     * Sends an email with a request to confirm some operation
     *
     * @param string $code      Operation activation code
     * @param array  $templates Templates to use in email
     *
     */
    protected function internalSendActivationEmail($code, $templates)
    {
        $user = $this;
        Mail::send(
            $templates,
            [
                'user_id' => $this->getId(),
                'code' => $code,
                'operation' => self::OPERATION_ACTIVATE_EMAIL
            ],
            function (Message $message) use ($user) {
                $from = Config::get('mail.from');
                if (empty($from)) {
                    throw new Exception("Email 'from' address is empty in mail config file");
                }
                $message->from($from['address'], $from['name']);
                $message->to($this->email);
                $message->subject(trans('email_subject_activate_email'));
            }
        );
    }
```

Variable $user in this example is detected as unused. But it is used. It is passed into the closure.

**Comment by niko2342** [comment]
+1

**Comment by nafigator** [comment]
> $message->to($this->email);

@FractalizeR did you mean this:

    $message->to($user->email);
? Is it work correctly with this fix?

**Comment by FractalizeR** [comment]
Oh... I was so long ago, but it looks fine now, thanks.

BTW, currently if I leave everything like in the snippet I posted (`$user` is passed into the closure, but its unused there), phpmd shows no warning about unused `$user` variable. Should it?

**Comment by nafigator** [comment]
> Should it?

@FractalizeR Looks like it really unused. But this depends on what is variable usage in phpmd. If a variable passage as parameter counts as usage, it should not.

**Comment by FractalizeR** [comment]
@nafigator Well, I think just passing parameter should not count as usage. Because... well, it is not usage, you know ;) Just passing.

**Comment by ravage84** [comment]
With PHPMD 2.6.0, in PhpStorm 2017.1.4:

``$user`` used normally, neither complains:
![unusedlocalvariableusednormally](https://cloud.githubusercontent.com/assets/625761/26749498/cf1feffc-480c-11e7-81e8-22a5bf5a6f4f.png)

``$user`` not used, PHPMD and PhpStorm display it as not used:
![unusedlocalvariablenotused](https://cloud.githubusercontent.com/assets/625761/26749509/08a44d18-480d-11e7-9f3c-bbf57fb78537.png)

``$user`` passed but not actually used, PHPMD complains, PhpStorm does not:
![unusedlocalvariablepassedtoclosrureonly](https://cloud.githubusercontent.com/assets/625761/26749489/9471b08e-480c-11e7-80ed-4cb64e7667d0.png)

So in this regard, PHPMD does not behave the same as PhpStorm, it's questionable whether it should, though.

Related: https://github.com/phpmd/phpmd/issues/325#issuecomment-297119165

Going to close this now. If anybody feels this needs to change, please open a new issue.
